packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol

```diff
@@ -158,4 +735,366 @@ library ConnextUtils {
   }
+
+  /**
+   * @notice Performs some sanity checks for `xcall`
+   * @dev Need this to prevent stack too deep
+   */
+  function _xcallSanityChecks(xCallLibArgs calldata _args) private {
+    // ensure this is the right domain
+    if (_args.xCallArgs.params.originDomain != _args.domain) {
+      revert ConnextUtils__xcall_wrongDomain();
+    }
+
+    // ensure theres a recipient defined
+    if (_args.xCallArgs.params.to == address(0)) {
+      revert ConnextUtils__xcall_emptyTo();
+    }
+  }
+
+  /**
+   * @notice Processes an `xcall`
+   * @dev Need this to prevent stack too deep
+   */
+  function _processXcall(
+    xCallLibArgs calldata _args,
+    mapping(address => ConnextMessage.TokenId) storage _adoptedToCanonical,
+    mapping(bytes32 => IStableSwap) storage _adoptedToLocalPools
+  )
+    private
+    returns (
+      bytes32,
+      bytes memory,
+      xCalledEventArgs memory
+    )
+  {
+    address transactingAssetId = _args.xCallArgs.transactingAssetId == address(0)
+      ? address(_args.wrapper)
+      : _args.xCallArgs.transactingAssetId;
+
+    // check that the asset is supported -- can be either adopted or local
+    ConnextMessage.TokenId memory canonical = _adoptedToCanonical[transactingAssetId];
+    if (canonical.id == bytes32(0)) {
+      revert ConnextUtils__xcall_notSupportedAsset();
+    }
+
+    // transfer funds of transacting asset to the contract from user
+    // NOTE: will wrap any native asset transferred to wrapped-native automatically
+    (, uint256 amount) = _handleIncomingAsset(_args);
+
+    // swap to the local asset from adopted
+    (uint256 bridgedAmt, address bridged) = _swapToLocalAssetIfNeeded(
+      canonical,
+      _adoptedToLocalPools[canonical.id],
+      _args.tokenRegistry,
+      transactingAssetId,
+      amount
+    );
+
+    bytes32 transferId = _getTransferId(_args, canonical);
+
+    bytes memory message = _formatMessage(_args, bridged, transferId, bridgedAmt);
+    _args.home.dispatch(_args.xCallArgs.params.destinationDomain, _args.remote, message);
+
+    return (
+      transferId,
+      message,
+      xCalledEventArgs({
+        transactingAssetId: transactingAssetId,
+        amount: amount,
+        bridgedAmt: bridgedAmt,
+        bridged: bridged
+      })
+    );
+  }
+
+  /**
+   * @notice Calculates a transferId base son `execute` arguments
+   * @dev Need this to prevent stack too deep
+   */
+  function _getTransferId(ExecuteLibArgs calldata _args) private view returns (bytes32) {
+    (uint32 tokenDomain, bytes32 tokenId) = _args.tokenRegistry.getTokenId(_args.executeArgs.local);
+
+    return
+      keccak256(
+        abi.encode(
+          _args.executeArgs.nonce,
+          _args.executeArgs.params,
+          _args.executeArgs.originSender,
+          tokenId,
+          tokenDomain,
+          _args.executeArgs.amount
+        )
+      );
+  }
+
+  /**
+   * @notice Calculates a transferId base son `xcall` arguments
```

Comment:
never talk to me or my base son ever again

Reply:
you've heard of JSON, now try base son

Reply:
:joy: